### ISSUE TITLE
Expose the match length when using String.IndexOf in culture specific mode

### ISSUE THREAD

**Issue by GSPP** [dotnet/corefx#33543]
`str.IndexOf(stringToFind)` returns the index of the start of the match but there is no way to tell the length of the match. The match length can be different from the search input (`stringToFind.Length`) in a culture specific mode such as `CurrentCulture`.

For example:

            Console.WriteLine("Straße".IndexOf("Straße", StringComparison.CurrentCulture)); //0
            Console.WriteLine("Straße".IndexOf("Strasse", StringComparison.CurrentCulture)); //0
            Console.WriteLine("Strasse".IndexOf("Straße", StringComparison.CurrentCulture)); //0
            Console.WriteLine("Strasse".IndexOf("Strasse", StringComparison.CurrentCulture)); //0

            Console.WriteLine("Straße".Length); //6
            Console.WriteLine("Straße".Normalize(NormalizationForm.FormC).Length); //6
            Console.WriteLine("Straße".Normalize(NormalizationForm.FormD).Length); //6
            Console.WriteLine("Strasse".Length); //7

`Straße` and `Strasse` are equivalent according to German string processing rules. The example code shows that the two forms match each other. This means that a longer string can be matched in a short string and the other way around.

The Framework should expose the length of the match. For example, in `"Straße".IndexOf("Strasse", StringComparison.CurrentCulture)` it should return `6` although `stringToFind.Length == 7`. In `"Strasse".IndexOf("Straße", StringComparison.CurrentCulture)` it should return `7` although `stringToFind.Length == 6`.

The motivating use case is HTML word highlighting. If I want to highlight all occurrences of `Straße` with `<b></b>` then I need to determine the length of the match.

Doing this currently requires Windows specific PInvoke or a search loop repeatedly calling `IndexOf`: https://stackoverflow.com/questions/20480016/length-of-substring-matched-by-culture-sensitive-string-indexof-method

I also found that `Regex` can accurately report the match length. But `Regex` is slow and awkward to use for non-constant search strings. I'm also not sure if `Regex` supports the exact same string processing semantics that the normal string functions have.

API proposal:

```
class String {
 public int IndexOf(string value, int startIndex, int count, StringComparison comparisonType, out int matchLength);
 public int LastIndexOf(string value, int startIndex, int count, StringComparison comparisonType, out int matchLength);
}
```

We could also add this to `StartsWith` and `EndsWith`. They are currently implemented in terms of `IndexOf`.

**Comment by joshfree** [comment]
@tarekgh @krwq

**Comment by tarekgh** [comment]
This proposal make sense and I believe we already have this internally implemented. 

@GSPP I think we need to add the APIs that takes spans too to this proposal.

**Comment by GSPP** [comment]
@tarekgh I have added the APIs you suggested plus StartsWith/EndsWith. Anything else that would make sense?

In my opinion we should *not* add overloads without `StringComparison comparisonType`. They would create a mess. This is an advanced scenario. Let's just have callers specify the comparison type they want.

**Comment by tarekgh** [comment]
Thanks @GSPP. it looks good to me. 

> In my opinion we should not add overloads without StringComparison comparisonType.

I agree with that and if anyone will need it later we can add it at that time.

**Comment by krwq** [comment]
Proposal looks good to me

**Comment by danmosemsft** [comment]
cc @benaadams 

I know the IndexOf methods are heavily optimized - to implement these, I wonder whether it would have measurable effects on perf - unless we duplicated all the code.

**Comment by tarekgh** [comment]
@danmosemsft the new APIs will just need plumbing as the code doing this functionality is there. I am not expecting this will affect the perf of other APIs.

**Comment by benaadams** [comment]
Not sure the non-ordinal versions (i.e culture) are greatly optimized; they just differ to the platform's implementation

/cc @GrabYourPitchforks

**Comment by GrabYourPitchforks** [comment]
@benaadams is correct. We do have some limited optimizations even for the culture-specific operations, but the APIs being proposed here aren't candidates for those optimizations.

@tarekgh, are you sure we have access to this information from our non-Windows globalization routines? I'm unfamiliar with those APIs, and from browsing the code it looks like some code paths (`LastIndexOf`, for instance) don't bubble this data up to the caller. I don't know if this is a limitation of the underlying libs or if it's something that can be corrected in our p/invoke.

**Comment by tarekgh** [comment]
> are you sure we have access to this information from our non-Windows globalization routines? 

Yes. here is the code https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs#L238 look at the last parameter which is the info we need :-) we already have it.

**Comment by GrabYourPitchforks** [comment]
@tarekgh Here's the actual p/invoke.

https://github.com/dotnet/coreclr/blob/98cd595b35ced3adb6efe0c667f5160f21067e0b/src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs#L291

In the `LastIndexOf` code path _specifically_, it's not flowed backward. This was the basis of my concern re: whether the underlying lib doesn't support it or if we just need to change our p/invoke.

**Comment by tarekgh** [comment]
@GrabYourPitchforks yes, that is the pluming I was referring to in my comment https://github.com/dotnet/corefx/issues/33543#issuecomment-469765841. we can use the existing pinvok with adding one extra parameter which I don't think will affect the perf. we'll just handle the extra parameter only in the new introduced cases to avoid affecting the existing APIs.

**Comment by terrajobst** [comment]
@GrabYourPitchforks mentioned that he's planning on addressing this issue in UTF8 string. He's thinking of exposing something like `TryFind` that outputs a `Range`. We should make sure the work aligns.
